File: ui/src/lib/api/messages.ts
import { apiGet, apiGetPage, apiPatch, apiPost } from "./client";

export type MessageKind = "dispatch" | "report" | "chat" | "status";

export interface Message {
  id: string;
  sender_kind: "agent" | "system" | "user";
  sender_agent_id: string | null;
  recipient_kind: "agent" | "user";
  recipient_agent_id: string | null;
  kind: MessageKind;
  subject: string;
  body: string;
  run_id: string | null;
  work_item_id: string | null;
  project_id: string | null;
  read_at: string | null;
  created_at: string;
}

export interface SendMessageInput {
  recipient_kind: "agent" | "user";
  recipient_agent_id?: string | null;
  kind?: MessageKind;
  subject?: string;
  body: string;
}

export const messageKeys = {
  list: (box: string) => ["messages", "list", box] as const,
  unread: (box: string) => ["messages", "unread", box] as const,
  sent: (agentId: string) => ["messages", "sent", agentId] as const,
};

export async function listMessages(box: string): Promise<Message[]> {
  const q = encodeURIComponent(box);
  return (await apiGetPage<Message[]>(`/messages?box=${q}&page_size=100`)).data;
}

export async function listSentMessages(agentId: string): Promise<Message[]> {
  const q = encodeURIComponent(agentId);
  return (await apiGetPage<Message[]>(`/messages?sender=${q}&page_size=100`)).data;
}

export async function getMessageUnreadCount(box: string): Promise<number> {
  const q = encodeURIComponent(box);
  return (await apiGet<{ count: number }>(`/messages/unread-count?box=${q}`)).count;
}

export async function markMessageRead(id: string): Promise<Message> {
  return apiPatch<Message>(`/messages/${id}`, { read: true });
}

export async function sendMessage(input: SendMessageInput): Promise<Message> {
  return apiPost<Message>("/messages", input);
}

File: src/interactors/api/routes/messages.py
from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel

from adapters.database.ports import UnitOfWork
from domain.base import utc_now
from domain.messages import Message, MessageKind, MessageRecipientKind, MessageSenderKind
from interactors.api.auth import current_user_id
from interactors.api.deps import get_uow
from interactors.api.envelope import ok

router = APIRouter(tags=["messages"])


def _box_filters(box: str | None) -> dict:
    """box='me' -> the user's mailbox; box=<agentId> -> that agent's mailbox."""
    if box == "me":
        return {"recipient_kind": "user"}
    if box:
        return {"recipient_kind": "agent", "recipient_agent_id": box}
    return {}


def _status_filters(status: str | None) -> dict:
    if status == "unread":
        return {"read_at__isnull": True}
    if status == "read":
        return {"read_at__isnull": False}
    return {}


@router.get("/messages")
def list_messages(
    box: str | None = Query(default=None),
    sender: str | None = Query(default=None),
    status: str | None = Query(default=None),
    page_size: int = Query(default=50),
    page_number: int = Query(default=1),
    uow: UnitOfWork = Depends(get_uow),
) -> dict:
    filters = {**_box_filters(box), **_status_filters(status)}
    if sender:
        filters["sender_agent_id"] = sender
    with uow.transaction():
        page = uow.messages.list(filters=filters, page_size=page_size,
                                 page_number=page_number, order_by="-created_at")
    return ok(
        [m.model_dump(mode="json") for m in page.results],
        meta={"total": page.total, "page_size": page.page_size, "page_number": page.page_number},
    )


@router.get("/messages/unread-count")
def unread_count(box: str | None = Query(default=None),
                 uow: UnitOfWork = Depends(get_uow)) -> dict:
    filters = {**_box_filters(box), "read_at__isnull": True}
    with uow.transaction():
        page = uow.messages.list(filters=filters, page_size=1)
    return ok({"count": page.total})


class PatchMessage(BaseModel):
    read: bool | None = None


@router.patch("/messages/{message_id}")
def patch_message(message_id: str, body: PatchMessage,
                  uow: UnitOfWork = Depends(get_uow)) -> dict:
    with uow.transaction():
        m = uow.messages.get(message_id)  # owner-scoped: cross-tenant -> 404
        updates: dict = {}
        if body.read and m.read_at is None:
            updates["read_at"] = utc_now()
        result = uow.messages.update(message_id, m.model_copy(update=updates))
    return ok(result.model_dump(mode="json"))


class SendMessage(BaseModel):
    recipient_kind: MessageRecipientKind
    recipient_agent_id: str | None = None
    kind: MessageKind = MessageKind.CHAT
    subject: str = ""
    body: str
    run_id: str | None = None
    work_item_id: str | None = None
    project_id: str | None = None


@router.post("/messages", status_code=201)
def send_message(body: SendMessage, user_id: str = Depends(current_user_id),
                 uow: UnitOfWork = Depends(get_uow)) -> dict:
    msg = Message(owner_id=user_id, sender_kind=MessageSenderKind.USER, **body.model_dump())
    with uow.transaction():
        created = uow.messages.create(msg)
    return ok(created.model_dump(mode="json"))
